src/statistical/unigram.jl

```diff
@@ -0,0 +1,177 @@
+    logp =  parse.(Float64, logprob)
+    spm = Sentencepiecemodel(vocab1, logp)
+    return spm
+end
+
+# to get index of given string
+function getindex(sp::Sentencepiecemodel, text)
+    id_list = findall(x->x==text, sp.vocab)
+    length(id_list) == 0 && return 2 #unk token index 
+    return id_list[1]
+end
+
+"""
+struct Nodes 
+    text::String
+    score::Float32
+    index::Int64
+    start::Int
+    en::Int
+end
+Utility structure, To hold the results of the forward pass (the forward Viterbi lattice)
+hold the token token string, score, vocabulary index, start and end character position   
+"""
+struct Nodes 
+    text::String
+    score::Float32
+    index::Int64
+    start::Int
+    en::Int
+end
+
+"""
+    decode_forward(sp::Sentencepiecemodel,text::String)
+Return all possible ngrams generated from sequence of items, as an Array{String,1}
+# Example
+```julia-repl
+julia> seq = ["To","be","or","not"]
+julia> a = everygram(seq,min_len=1, max_len=-1)
+ 10-element Array{Any,1}:
+  "or"          
+  "not"         
+  "To"          
+  "be"                  
+  "or not" 
+  "be or"       
+  "be or not"   
+  "To be or"    
+  "To be or not"
+``` 
+"""
+function decode_forward(sp::Sentencepiecemodel, text::String)
+    results = Array{Nodes, 1}(undef, length(text))
+    scores = fill(-Inf, length(text))
+    scores[1] =0
+    for char_end in 1:length(text)
+        for char_start in 1:char_end
+            if text[char_start:char_end] in sp.vocab
+                subtokenid = getindex(sp, text[char_start:char_end])[1]
+                local_score = scores[char_start] + sp.logprob[subtokenid]
+                if local_score > scores[char_end]   
+                    results[char_end] = Nodes(text[char_start:char_end], local_score, subtokenid, char_start, char_end)
+                    scores[char_end] = local_score
+                end
+            end
+        end
+        if scores[char_end] == -Inf
+            results[char_end] = Nodes(text[char_end-1:char_end], -Inf, 1, char_end-1, char_end)
+            scores[char_end] = 0
+        end
+        if scores[char_end] == 0
+            results[char_end] = Nodes(text[char_end:char_end], -Inf, 1, char_end, char_end)
+        end
+    end
+    return(results)
+end
+
+"""
+    decode_forward(sp::Sentencepiecemodel,text::String)
+Return all possible ngrams generated from sequence of items, as an Array{String,1}
+"""
+function decode_backward(sp::Sentencepiecemodel, nodes)
+    next_nodes = nodes[end]
+    best_seq = []
+    
+    while next_nodes.start > 1
+        node_value = next_nodes
+        next_nodes = nodes[(node_value.start)-1]
+        push!(best_seq, node_value)
+    end
+    push!(best_seq, next_nodes)
+    return(best_seq)
+end
+
+"""
+    Tokenizer(sp::Sentencepiecemodel,text::AbstractString)
+given spm path and text it tokenized you string
+It does all the preprocessing step needed 
+"""
+function tokenizer(sp::Sentencepiecemodel, text::AbstractString)
```

Comment:
How about changing this to `function (sp::Sentencepiecemodel)(text:AbstractString)`
This way instead of `tokenizer(spm, "Some string here.")` we can use `spm("Some string here.")`

Reply:
sure I am adding both the methods